src/main/java/com/feedzai/commons/sql/abstraction/engine/AbstractDatabaseEngine.java

```diff
@@ -1103,6 +1104,88 @@ private void validateEntity(final DbEntity entity) throws DatabaseEngineExceptio
      */
     protected abstract boolean checkConnection(final Connection conn);
 
+    /**
+     * Executes a dummy query to verify if the connection is alive.
+     * @implNote This method should be engine aware since some engines use different approaches.
+     *
+     * @param conn The connection to test.
+     * @return True if the connection is valid, false otherwise.
+     * @since 2.7.2
+     */
+    protected boolean pingConnection(final Connection conn) {
+        Statement s = null;
+        try {
+            s = conn.createStatement();
+            s.executeQuery("select 1");
+
+            return true;
+        } catch (final SQLException e) {
+            logger.debug("Connection is down.", e);
+            return false;
+        } finally {
+            try {
+                if (s != null) {
+                    s.close();
+                }
+            } catch (final Exception e) {
+                logger.trace("Error closing statement.", e);
+            }
+        }
+    }
+
+    /**
+     * Checks if the connection is alive with a given timeout.
+     * @implNote Engine aware implementations are recommended since there are some engines that do not handle
+     * {@link Connection#setNetworkTimeout(Executor, int)} synchronously, throwing unexpected exceptions, e.g:
+     * <ul>
+     *     <li>Oracle</li>
+     *     <li>MySQL</li>
+     *</ul>
+     * Also, engines that accept correctly the {@link Connection#isValid(int)} it is recommented to use e.g:
+     * <ul>
+     *     <li>Oracle</li>
+     *     <li>MySQL</li>
+     *     <li>SQLServer</li>
+     *     <li>H2</li>
+     *     <li>DB2</li>
+     * </ul>
+     * PostgreSQL is not respecting the timeout in the {@link Connection#isValid(int)}.
+     *
+     * @param conn    The connection to test.
+     * @param timeout The timeout in seconds.
+     * @return {@code true} if the connection is valid, {@code false} otherwise.
+     * @since 2.7.2
+     */
+    protected boolean checkConnection(final Connection conn, final int timeout) {
```

Comment:
Sorry but if this checkConnection is intended to be used only in the Postgres implementation, why don't we move this method to that specific implementation and move here the method that does the #isValid check (we can then remove the checkConnection from the other implementations).

I would also make the #pingConnection visible only in the Postgres class since it just needs to be used on that implementation.

Do you agree?